### PR TITLE
Refactor CardReaderOnboardingChecker to allow user to choose between multiple plugins

### LINKED ISSUE #6626: Update OnboardingChecker - Clear selected payment gateway if onboarding doesn't complete successfully
Goal of this task is to clear selected payment gateway if onboarding doesn't complete successfully. The intention here is to make sure the user has an option to change their selection in case the onboarding doesn't complete successfully. Only when FF is enabled.

### PR DESCRIPTION
**Pull request by AnirudhBhat**
Closes: #6626 

### Description

This PR does the following:
1. Introduce the IPP Payment Gateway feature flag into the `CardReaderOnboardingChecker` class
2. Inject `PluginType` into `CardReaderOnboardingChecker` class
3. Updates the "isPluginExplicitlySelected" flag to `false` when the onboarding fails for any reason.
4. Adds a new state `ChoosePaymentGatewayProvider`
5. Returns `ChoosePaymentGatewayProvider` when the IPP Payment Gateway feature flag is enabled and the "isPluginExplicitlySelected" flag is set to `false` and user hasn't selected the plugin yet.

### Testing instructions

ViewModel and the UI aren't connected to this logic yet. Green CI should be enough.

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Review by rossanafmenezes — APPROVED** [review]
LGTM 👍 I had a few questions, but feel free to merge it :)

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingChecker.kt [hunk 1]
```diff
@@ -121,10 +129,30 @@ class CardReaderOnboardingChecker @Inject constructor(
                     return PluginIsNotSupportedInTheCountry(pluginType, countryCode!!)
                 }
             }
-            return WcpayAndStripeActivated
+            if (ippSelectPaymentGateway.isEnabled()) {
+                when {
+                    isUserComingFromChoosePaymentGatewayScreen(pluginType) -> {
+                        updateSharedPreferences(
```
Comment (rossanafmenezes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingChecker.kt:135]: ❓ Could you please elaborate a bit on why do we need to store the information in the SharedPreferences here and not rely on it being stored on the code on the lines 85 to 94?

```
val (status, version) = when (it) {
                    is OnboardingCompleted -> CARD_READER_ONBOARDING_COMPLETED to it.version
                    is StripeAccountPendingRequirement -> CARD_READER_ONBOARDING_PENDING to it.version
                    else -> {
                        if (ippSelectPaymentGateway.isEnabled()) {
                            updatePluginExplicitlySelectedFlag(false)
                        }
                        CARD_READER_ONBOARDING_NOT_COMPLETED to null
                    }
                }
```

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingFragment.kt [hunk 2]
```diff
@@ -264,7 +264,10 @@ sealed class CardReaderOnboardingParams : Parcelable {
     abstract val cardReaderFlowParam: CardReaderFlowParam
 
     @Parcelize
-    data class Check(override val cardReaderFlowParam: CardReaderFlowParam) : CardReaderOnboardingParams()
+    data class Check(
+        override val cardReaderFlowParam: CardReaderFlowParam,
+        val pluginType: PluginType? = null
```
Comment (rossanafmenezes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingFragment.kt:269]: ❓ Could you please clarify why we have this param here? I could not find usage for it. 🤔

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingCheckerTest.kt [hunk 3]
```diff
@@ -1115,6 +1120,343 @@ class CardReaderOnboardingCheckerTest : BaseUnitTest() {
...
+                )
+            ).thenReturn(false)
+
+            val result = checker.getOnboardingState()
+
+            assertThat(result).isEqualTo(CardReaderOnboardingState.ChoosePaymentGatewayProvider)
+        }
+
+    @Test
+    fun `given plugin selected, when multiple plugins, then set plugin flag`() =
```
Comment (rossanafmenezes) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingCheckerTest.kt:1268]: 🔍 Feels like the "given" and "when" are inverted here. Just wondering if shouldn't it be Given multiple plugins activated when user selects plugin... wdyt?

**Review by rossanafmenezes — APPROVED** [review]

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingCheckerTest.kt [hunk 4]
```diff
@@ -1115,6 +1120,343 @@ class CardReaderOnboardingCheckerTest : BaseUnitTest() {
...
+
+            verify(appPrefsWrapper).getCardReaderPreferredPlugin(
+                anyInt(),
+                anyLong(),
+                anyLong(),
+            )
+        }
+
+    @Test
+    fun `when navigating to onboarding via payment gateway screen, then store the plugin to the shared preference`() =
```
Comment (malinajirka) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingCheckerTest.kt:1392]: 1.  🔍 I'm wondering if the test actually tests what the name suggests. The test tests that when `WCPay` is passed as the selected plugin and `isCardReaderPluginExplicitlySelected` is true -> the checker stores `CARD_READER_ONBOARDING_NOT_COMPLETED` and `WOOCOMMERCE_PAYMENTS`. Right? (Isn't this state unreachable? How come we are passing a selected plugin and the `isExplicitlySelected` is already true. This shouldn't happen in production, right?)

3. ⚠️ Btw is this the behavior we want? Shouldn't the onboarding checker clear the selection when onboarding is not completed? Why do we need to know which plugin the user selected when the onboarding fails? Won't the user get stuck in the following scenario:
```
1. WCPay is preferred and is explicitly selected
2. User clicks on "Manage Payment Gateway"
5. User selects Stripe Extension
6. The Stripe extension version is not supported -> onboarding error is shown
7. The onboarding is now stuck forever on the "Unsupported plugin version" screen since the app has `Stripe Extension` stored as the preferred plugin.
```
If this issue is not present because we clear the `isExplicitlySelected` flag then I guess we are fine, but I still don't understand why we need to store the plugin they selected.

🔍 nitpick -> if we need to store the state so the other parts of the onboarding checker read it from shared prefs within the same run -> it feels a bit off that the onboarding checker needs to store something into a persistent storage so it can read it a few microseconds after (perhaps, it could use a local variable or the passed "pluginType" parameter), wdyt? 

Sorry, I realize this comment might be difficult to grasp - It can be summarised to the following questions:
- How come we are passing a selected plugin into the onboarding checker and the `isExplicitlySelected` is already true. This shouldn't happen in production, right?
- Why do we need to store which plugin the user selected when the onboarding fails?
- Does the onboarding checker read a value from shared prefs which was saved during the same run of the onboarding checker, if so, is that really needed - can't we use a local variable or the `pluginType` parameter?


Thanks @AnirudhBhat !

Comment (AnirudhBhat) [reply WooCommerce/src/test/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingCheckerTest.kt:1392]: >I'm wondering if the test actually tests what the name suggests. The test tests that when WCPay is passed as the selected plugin and isCardReaderPluginExplicitlySelected is true -> the checker stores CARD_READER_ONBOARDING_NOT_COMPLETED and WOOCOMMERCE_PAYMENTS. Right? (Isn't this state unreachable? How come we are passing a selected plugin and the isExplicitlySelected is already true. This shouldn't happen in production, right?)

Here's what is happening
1. `isExplicitlySelected` flag is false and multiple plugins detected
2. Merchant selects WCPay as their plugin and clicks continue (Note that we are not setting `isExplicitlySelected` flag to true here yet!)
3. We restart the onboarding screen
4. WCPay is passed as a parameter to the onboarding
5. Onboarding again detects that multiple plugins are installed
6. Since the onboarding class sees "Plugin type" in the parameter, it knows the merchant is coming from the "Choose payment gateway" screen
7. At this point, `isExplicitlySelected` flag is still false. 
8. We now know for sure the plugin is explicitly selected by the merchant. We set `isExplicitlySelected` flag to true (At this point, we have a plugin type and `isExplicitlySelected` is true)
9. We also update the shared preference with the user-selected plugin

Regarding point 8. It happens (having both plugin type and `isExplicitlySelected` true) only when the user selects any of the plugins from the "Choose payment gateway" screen and navigates to the onboarding checker screen as a result. But please note that even in this case, `isExplicitlySelected` will be false initially when the onboarding checker runs. Later when the onboarding checker detects that there is some plugin type selected, we switch the flag to true. 

Reason for not setting the `isExplicitlySelected` flag to true as soon as we detect multiple plugins and show the "Choose payment gateway" screen is because there are chances that the merchant can bail out of that screen without selecting anything. In this case, we end up with an invalid state.


Regarding point 9 -> Why we store merchant selected plugin to the shared preference is because regardless of from where the merchant is coming to the onboarding screen (settings, payment gateway screen...etc), if `isExplicitlySelected` is true. We need to fetch the plugin from the shared preference because merchant had already pre-selected their desired plugin.

The only reason why I stored merchant selected plugin into the shared preference when we detect the merchant is coming to the onboarding screen via the "choose payment gateway" screen is that I wanted to make it less confusing the readers. In one case, we use the plugin type as the desired plugin and in another case, we use the value from the shared preference as the desired plugin so that we could avoid one more `if else` condition. 

But it turns out it is in fact causing more confusion 😄 

>can't we use a local variable or the pluginType parameter?

Yes! We can. We can make use of the plugin type parameter for detecting the merchant selected plugin only in the case when the merchant lands on the onboarding screen as a result of choosing any one of the plugins in the previous screen. Only in this case, the plugin type will not be null.

However, if the merchant had already selected the plugin and is navigating to the onboarding screen via the IPP settings screen. The plugin type will be null and we need to fetch the merchant's pre-selected plugin from the shared preference. 

But yeah, we could definitely make use of the plugin type instead of querying shared preference when the merchant lands on the onboarding screen as a result of choosing any one of the plugins in the previous screen.


>  WCPay is preferred and is explicitly selected
User clicks on "Manage Payment Gateway"
User selects Stripe Extension
The Stripe extension version is not supported -> onboarding error is shown
The onboarding is now stuck forever on the "Unsupported plugin version" screen since the app has `Stripe Extension` stored as the preferred plugin.


1. WCPay is preferred and is explicitly selected (`isExplicitlySelected` will be true)
2. User clicks on "Manage Payment Gateway" (we clear `isExplicitlySelected` flag)
3. User selects Stripe Extension
4. The Stripe extension version is not supported -> onboarding error is shown (we clear `isExplicitlySelected` flag)
5. The onboarding is now stuck forever on the "Unsupported plugin version" screen since the app has `Stripe Extension` stored as the preferred plugin.

On the 5th step, the merchant enters the onboarding screen again:
1. We detect multiple gateways
2. We detect that the `isEplicitlySelected` flag is false
3. We show them "Choose payment gateway" screen.
4. Let's assume Merchant selects WCPay
5. The cycle repeats

I don't think the merchant gets stuck anywhere. WDYT?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingFragment.kt (hunk 2 above)
Comment (AnirudhBhat) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingFragment.kt:269]: This is the param that indicates the plugin type the merchant selected in the "Choose payment gateway" screen. Right now, it's used in `CardReaderOnboardingViewModel` line no: `90`

Whenever we call the onboarding screen from anywhere in the app, we pass it a param called `CardReaderOnboardingParam`

`CardReaderOnboardingParam` is a sealed class that has two subtypes. `Check` and `Failed`

We are passing the plugin type as a parameter when we detect the `CardReaderOnboardingParam` is of type  `Check`

Basically, when the merchant selects a plugin. `CardReaderOnboardingViewModel` knows which plugin type the merchant selected since that's where this logic lives (This logic isn't implemented yet). `CardReaderOnboardingViewModel` restarts the onboarding screen passing in this plugin type as a param.

Let me know if I'm making sense

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingChecker.kt (hunk 1 above)
Comment (AnirudhBhat) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingChecker.kt:135]: Good question! the code on the lines 85 to 94 stores onboarding information into the shared preference when the onboarding screen returns due to onboarding being successful or failure. However, we are storing here the merchant-selected plugin info when the merchant lands on the onboarding screen via choose payment gateway screen. In this case, We can't use the stored information since the merchant has just selected the plugin and we have restarted the onboarding screen and we haven't stored the information yet.

We can remove storing here and instead just use plugin type passed as param instead for the case when the merchant lands on the onboarding screen from the "choose payment gateway" screen. For all other cases like merchant navigating to the onboarding screen via IPP settings, we can just refer to the already stored values for the plugin type. 

You can see more about this here: https://github.com/woocommerce/woocommerce-android/pull/6796#discussion_r910584524

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingCheckerTest.kt (hunk 4 above)
Comment (malinajirka) [reply WooCommerce/src/test/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingCheckerTest.kt:1392]: Thanks so much for taking the time to reply!
> 1. isExplicitlySelected flag is false and multiple plugins detected 
> Regarding point 8. It happens (having both plugin type and isExplicitlySelected true) only when the user selects any of the plugins from the "Choose payment gateway" screen and navigates to the onboarding checker screen as a result. But please note that even in this case, isExplicitlySelected will be false initially when the onboarding checker runs. Later when the onboarding checker detects that there is some plugin type selected, we switch the flag to true.

That was my understanding too, however the test contains 
```
 whenever(
                appPrefsWrapper.isCardReaderPluginExplicitlySelected(
                    anyInt(),
                    anyLong(),
                    anyLong(),
                )
            ).thenReturn(true)
```
so the flag is `true` before the test starts, right (which doesn't correspond to neither step 1 nor step 4 in your example)? So it feels like the test is testing a scenario which can't happen in production. Wdyt?

> Reason for not setting the isExplicitlySelected flag to true as soon as we detect multiple plugins and show the "Choose payment gateway" screen is because there are chances that the merchant can bail out of that screen without selecting anything. In this case, we end up with an invalid state.

Yep this is clear, thanks!

> But yeah, we could definitely make use of the plugin type instead of querying shared preference when the merchant lands on the onboarding screen as a result of choosing any one of the plugins in the previous screen.

I feel like "Did someone passed us a plugin they want to use, let's use it otherwise load it from storage" is easier to reason about. However, this is about a personal preference, so no need to change anything.

> 1. We detect multiple gateways
> 2. We detect that the isEplicitlySelected flag is false
> 3. We show them "Choose payment gateway" screen.
> 4. Let's assume Merchant selects WCPay
> 5. The cycle repeats

Thanks for the clarification, it seems this explanation corresponds to the other case I mentioned:  `If this issue is not present because we clear the isExplicitlySelected flag then I guess we are fine...`

I see where my confusion is coming from, it's related to [the other discussion](https://github.com/woocommerce/woocommerce-android/pull/6796#discussion_r910639973) -> I'll reply there.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingFragment.kt (hunk 2 above)
Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingFragment.kt:269]: Rosie is AFK and I was pair reviewing this PR with here, so I'll reply to speed things up (hopefully, she won't mind 🤞).

I believe I understand the logic and it makes sense. However, there is one thing which is still a bit unclear to me.

We want to pass the `pluginType` only after the app displayed the "Select Payment Gateway" screen and the user selected a plugin, right?

The only fragment which is capable of displaying "Select Payment Gateway" is CardReaderOnboardingFragment, correct?

If the above is correct, wouldn't the user be already on the CardReaderOnboardingFragment/ViewModel when they select a plugin -> can't the fragment/viewmodel just call `refreshState` instead of navigating to itself again and passing the selectedPlugin as parameter? Wdyt @AnirudhBhat?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingChecker.kt (hunk 1 above)
Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingChecker.kt:135]: Rosie is AFK and I was pair reviewing this PR with here, so I'll reply to speed things up (hopefully, she won't mind 🤞).

Thanks @AnirudhBhat!

If I understand this correctly, your are saying that we could:
1. Remove invocation of `updateSharedPreferences` from line 135.
2. Use the passed pluginType parameter for the case when the merchant lands on the onboarding screen from the "choose payment gateway" screen.

Is this the same thing as saying the following?
1. Remove invocation of `updateSharedPreferences` from line 135.
2. No matter the case/scenario, use `pluginType` param when it's not null, otherwise use the stored value.

Or is there a difference? If there is no difference I'd personal vote for modifying the code as it feels simpler. Whoever reads the code of the onboarding checker might not have all the context which scenarios are possible and which are not (the checker shouldn't care either) -> reasoning about "someone passed me a value so I use it otherwise I load it from shared prefs" feels easier. Having said that, it's just my opinion so no need to change anything.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingFragment.kt (hunk 2 above)
Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingFragment.kt:269]: I was thinking about this and I believe we can keep the code there and if we find that we don't need it when the whole feature is implemented, we can simply refactor it and remove it.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingChecker.kt (hunk 1 above)
Comment (malinajirka) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/cardreader/onboarding/CardReaderOnboardingChecker.kt:135]: This feels like as a potential improvement (if I'm not missing anything), but doesn't feel like something we should block this PR on. Also the current code works and is already tested so I think the effort of refactoring it might not be worth it.

**Review by malinajirka — APPROVED** [review]
Thanks so much for implementing this @AnirudhBhat! I gave my review comments more thoughts and I realized none of them was blocking and they might not be worth the effort. Since Rosie approved the PR it seems like she didn't consider her comments blocking either, so I believe we are good to merge. 🥳
